sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -37,51 +37,56 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
-    // not support columnar, a projection is added to convert the rows to UnsafeRow.
-    case PhysicalOperation(project, filters, relation: DataSourceV2ScanRelation) =>
-      val output = relation.output
+    case PhysicalOperation(project, filters,
+        relation @ DataSourceV2ScanRelation(table, v1Scan: V1Scan, output)) =>
       val pushedFilters = relation.getTagValue(V2ScanRelationPushDown.PUSHED_FILTERS_TAG)
         .getOrElse(Seq.empty)
-
-      val (scanExec, needsUnsafeConversion) = relation.scan match {
-        case v1Scan: V1Scan =>
-          val v1Relation = v1Scan.toV1TableScan(session.sqlContext)
-          if (v1Relation.schema != v1Scan.readSchema()) {
-            throw new IllegalArgumentException(
-              "The fallback v1 relation reports inconsistent schema:\n" +
-                "Schema of v2 scan:     " + v1Scan.readSchema() + "\n" +
-                "Schema of v1 relation: " + v1Relation.schema)
-          }
-          val rdd = v1Relation match {
-            case s: TableScan => s.buildScan()
-            case _ =>
-              throw new IllegalArgumentException(
-                "`V1Scan.toV1Relation` must return a `TableScan` instance.")
-          }
-          val unsafeRowRDD = DataSourceStrategy.toCatalystRDD(v1Relation, output, rdd)
-          val originalOutputNames = relation.table.schema().map(_.name)
-          val requiredColumnsIndex = output.map(_.name).map(originalOutputNames.indexOf)
-          val dsScan = RowDataSourceScanExec(
-            output,
-            requiredColumnsIndex,
-            pushedFilters.toSet,
-            pushedFilters.toSet,
-            unsafeRowRDD,
-            v1Relation,
-            tableIdentifier = None)
-          (dsScan, false)
+      val v1Relation = v1Scan.toV1TableScan(session.sqlContext)
+      if (v1Relation.schema != v1Scan.readSchema()) {
+        throw new IllegalArgumentException(
+          "The fallback v1 relation reports inconsistent schema:\n" +
+            "Schema of v2 scan:     " + v1Scan.readSchema() + "\n" +
+            "Schema of v1 relation: " + v1Relation.schema)
+      }
+      val rdd = v1Relation match {
+        case s: TableScan => s.buildScan()
         case _ =>
-          val batchScan = BatchScanExec(output, relation.scan)
-          (batchScan, !batchScan.supportsColumnar)
+          throw new IllegalArgumentException(
+            "`V1Scan.toV1Relation` must return a `TableScan` instance.")
       }
+      val unsafeRowRDD = DataSourceStrategy.toCatalystRDD(v1Relation, output, rdd)
+      val originalOutputNames = relation.table.schema().map(_.name)
+      val requiredColumnsIndex = output.map(_.name).map(originalOutputNames.indexOf)
+      val dsScan = RowDataSourceScanExec(
+        output,
+        requiredColumnsIndex,
+        pushedFilters.toSet,
+        pushedFilters.toSet,
+        unsafeRowRDD,
+        v1Relation,
+        tableIdentifier = None)
+
+      val filterCondition = filters.reduceLeftOption(And)
```

Comment:
how about abstracting
```
     val filterCondition = filters.reduceLeftOption(And)
      val withFilter = filterCondition.map(FilterExec(_, dsScan)).getOrElse(dsScan)


      val withProjection = if (withFilter.output != project) {
        ProjectExec(project, withFilter)
      } else {
        withFilter
      }

      withProjection :: Nil
```
to one method `withProjectAndFilter(project, filter, leaf)`